flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java

```diff
@@ -637,9 +675,23 @@ private boolean isPatternOptional(Pattern<T, ?> pattern) {
 				untilCondition,
 				true);
 
-			final IterativeCondition<T> proceedCondition = getTrueFunction();
+			IterativeCondition<T> proceedCondition = getTrueFunction();
 			final State<T> loopingState = createState(currentPattern.getName(), State.StateType.Normal);
-			loopingState.addProceed(sinkState, proceedCondition);
+
+			if (currentPattern.getQuantifier().hasProperty(Quantifier.QuantifierProperty.GREEDY)) {
+				State<T> sinkStateCopy = copy(sinkState);
```

Comment:
If I understood correctly the copies are needed only in casese where there is the `untilCondition`. Am I right? If so let's create the copy then. Right know there are dangling copies of the next when there is no `untilCondtion`.